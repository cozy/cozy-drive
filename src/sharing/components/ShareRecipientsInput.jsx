import React, { Component } from 'react'
import PropTypes from 'prop-types'

import ShareAutosuggest from 'sharing/components/ShareAutosuggest'
import styles from 'sharing/share.styl'

class ShareRecipientsInput extends Component {
  static contextTypes = {
    client: PropTypes.object.isRequired
  }

  state = {
    loading: false
  }

  componentDidUpdate() {
    const { contacts, groups } = this.props
    if (
      this.state.loading &&
      !contacts.hasMore &&
      contacts.fetchStatus === 'loaded' &&
      !groups.hasMore &&
      groups.fetchStatus === 'loaded'
    ) {
      this.setState({
        loading: false
      })
    }
  }

  onFocus = () => {
    const { contacts, groups } = this.props
    if (
      contacts.hasMore ||
      contacts.fetchStatus === 'loading' ||
      groups.hasMore ||
      groups.fetchStatus === 'loading'
    ) {
      this.setState({
        loading: true
      })
    }
  }

  getContactsAndGroups = () => {
    // we need contacts to be loaded to be able to add all group members to recipients
    const { contacts, groups } = this.props
    if (contacts.hasMore || contacts.fetchStatus === 'loading') {
      return contacts.data
    } else {
      return [...contacts.data, ...groups.data]
    }
  }

  render() {
    const { label, onPick, onRemove, placeholder, recipients } = this.props
    const { loading } = this.state
    return (
      <div>
        <label className={styles['coz-form-label']} htmlFor="email">
          {label}
        </label>
        <ShareAutosuggest
          loading={loading}
          contactsAndGroups={this.getContactsAndGroups()}
          recipients={recipients}
          onFocus={this.onFocus}
          onPick={onPick}
          onRemove={onRemove}
          placeholder={placeholder}
        />
      </div>
    )
  }
}

ShareRecipientsInput.propTypes = {
  label: PropTypes.string,
  contacts: PropTypes.object.isRequired,
  groups: PropTypes.object.isRequired,
  recipients: PropTypes.array,
  onPick: PropTypes.func.isRequired,
  onRemove: PropTypes.func.isRequired,
  placeholder: PropTypes.string
}

ShareRecipientsInput.defaultProps = {
  label: 'To:',
  contacts: [],
  recipients: []
}

export default ShareRecipientsInput
