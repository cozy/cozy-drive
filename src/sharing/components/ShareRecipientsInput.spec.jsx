import React from 'react'
import { shallow } from 'enzyme'

import ShareRecipientsInput from './ShareRecipientsInput'

describe('ShareRecipientsInput component', () => {
  it('should match snapshot', () => {
    const props = {
      label: 'Recipients',
      contacts: {
        id: 'contacts',
        fetchStatus: 'loaded',
        hasMore: false,
        data: [
          {
            id: 'df563cc4-6440',
            name: {
              givenName: 'Michale',
              familyName: 'Russel'
            }
          },
          {
            id: '5a3b4ccf-c257',
            name: {
              givenName: 'Teagan',
              familyName: 'Wolf'
            }
          }
        ]
      },
      groups: {
        id: 'groups',
        fetchStatus: 'loaded',
        hasMore: false,
        data: [
          { id: 'fe86af20-c6c5', name: "The Night's Watch" },
          { id: '3d8193ab-2ce4', name: 'The North' }
        ]
      },
      recipients: [
        {
          id: '5a3b4ccf-c257',
          name: {
            givenName: 'Teagan',
            familyName: 'Wolf'
          }
        }
      ],
      onPick: jest.fn().mockName('onPick'),
      onRemove: jest.fn().mockName('onRemove'),
      placeholder: 'Enter recipients here'
    }
    const jsx = <ShareRecipientsInput {...props} />
    const wrapper = shallow(jsx)
    expect(wrapper).toMatchSnapshot()
  })

  it('should include groups only if all contacts are loaded', () => {
    const props = {
      label: 'Recipients',
      contacts: {
        id: 'contacts',
        fetchStatus: 'loading',
        hasMore: false,
        data: [
          {
            id: 'df563cc4-6440',
            name: {
              givenName: 'Michale',
              familyName: 'Russel'
            }
          },
          {
            id: '5a3b4ccf-c257',
            name: {
              givenName: 'Teagan',
              familyName: 'Wolf'
            }
          }
        ]
      },
      groups: {
        id: 'groups',
        fetchStatus: 'loaded',
        hasMore: false,
        data: [
          { id: 'fe86af20-c6c5', name: "The Night's Watch" },
          { id: '3d8193ab-2ce4', name: 'The North' }
        ]
      },
      recipients: [
        {
          id: '5a3b4ccf-c257',
          name: {
            givenName: 'Teagan',
            familyName: 'Wolf'
          }
        }
      ],
      onPick: jest.fn().mockName('onPick'),
      onRemove: jest.fn().mockName('onRemove'),
      placeholder: 'Enter recipients here'
    }
    const jsx = <ShareRecipientsInput {...props} />
    const wrapper = shallow(jsx)
    expect(wrapper).toMatchSnapshot()
  })
})
