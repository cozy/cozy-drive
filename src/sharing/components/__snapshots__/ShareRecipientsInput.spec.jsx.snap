// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareRecipientsInput component should include groups only if all contacts are loaded 1`] = `
<div>
  <label
    className="coz-form-label"
    htmlFor="email"
  >
    Recipients
  </label>
  <ShareAutocomplete
    contactsAndGroups={
      Array [
        Object {
          "id": "df563cc4-6440",
          "name": Object {
            "familyName": "Russel",
            "givenName": "Michale",
          },
        },
        Object {
          "id": "5a3b4ccf-c257",
          "name": Object {
            "familyName": "Wolf",
            "givenName": "Teagan",
          },
        },
      ]
    }
    loading={false}
    onFocus={[Function]}
    onPick={[MockFunction onPick]}
    onRemove={[MockFunction onRemove]}
    placeholder="Enter recipients here"
    recipients={
      Array [
        Object {
          "id": "5a3b4ccf-c257",
          "name": Object {
            "familyName": "Wolf",
            "givenName": "Teagan",
          },
        },
      ]
    }
  />
</div>
`;

exports[`ShareRecipientsInput component should match snapshot 1`] = `
<div>
  <label
    className="coz-form-label"
    htmlFor="email"
  >
    Recipients
  </label>
  <ShareAutocomplete
    contactsAndGroups={
      Array [
        Object {
          "id": "df563cc4-6440",
          "name": Object {
            "familyName": "Russel",
            "givenName": "Michale",
          },
        },
        Object {
          "id": "5a3b4ccf-c257",
          "name": Object {
            "familyName": "Wolf",
            "givenName": "Teagan",
          },
        },
        Object {
          "id": "fe86af20-c6c5",
          "name": "The Night's Watch",
        },
        Object {
          "id": "3d8193ab-2ce4",
          "name": "The North",
        },
      ]
    }
    loading={false}
    onFocus={[Function]}
    onPick={[MockFunction onPick]}
    onRemove={[MockFunction onRemove]}
    placeholder="Enter recipients here"
    recipients={
      Array [
        Object {
          "id": "5a3b4ccf-c257",
          "name": Object {
            "familyName": "Wolf",
            "givenName": "Teagan",
          },
        },
      ]
    }
  />
</div>
`;
